pkg/ruler/manager/memstore.go

```diff
@@ -0,0 +1,380 @@
+)
+
+const (
+	AlertForStateMetricName = "ALERTS_FOR_STATE"
+	statusSuccess           = "success"
+	statusFailure           = "failure"
+)
+
+type NoopAppender struct{}
+
+func (a NoopAppender) Appender() storage.Appender                              { return a }
+func (a NoopAppender) Add(l labels.Labels, t int64, v float64) (uint64, error) { return 0, nil }
+func (a NoopAppender) AddFast(ref uint64, t int64, v float64) error {
+	return errors.New("unimplemented")
+}
+func (a NoopAppender) Commit() error   { return nil }
+func (a NoopAppender) Rollback() error { return nil }
+
+func ForStateMetric(base labels.Labels, alertName string) labels.Labels {
+	b := labels.NewBuilder(base)
+	b.Set(labels.MetricName, AlertForStateMetricName)
+	b.Set(labels.AlertName, alertName)
+	return b.Labels()
+}
+
+type Metrics struct {
+	Evaluations *prometheus.CounterVec
+	Samples     prometheus.Gauge       // in memory samples
+	CacheHits   *prometheus.CounterVec // cache hits on in memory samples
+}
+
+func NewMetrics(r prometheus.Registerer) *Metrics {
+	return &Metrics{
+		Evaluations: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "ruler_memory_for_state_evaluations",
+		}, []string{"status", "tenant"}),
+		Samples: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Namespace: "loki",
+			Name:      "ruler_memory_samples",
+		}),
+		CacheHits: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "ruler_memory_for_state_cache_hits",
+		}, []string{"tenant"}),
+	}
+}
+
+type RuleIter interface {
+	AlertingRules() []*rules.AlertingRule
+}
+
+type MemStore struct {
+	mtx       sync.Mutex
+	userID    string
+	queryFunc rules.QueryFunc
+	metrics   *Metrics
+	mgr       RuleIter
+	logger    log.Logger
+	rules     map[string]*RuleCache
+
+	initiated       chan struct{}
+	done            chan struct{}
+	cleanupInterval time.Duration
+}
+
+func NewMemStore(userID string, queryFunc rules.QueryFunc, metrics *Metrics, cleanupInterval time.Duration, logger log.Logger) *MemStore {
+	s := &MemStore{
+		userID:          userID,
+		metrics:         metrics,
+		queryFunc:       queryFunc,
+		logger:          logger,
+		cleanupInterval: cleanupInterval,
+		rules:           make(map[string]*RuleCache),
+
+		initiated: make(chan struct{}), // blocks execution until Start() is called
+		done:      make(chan struct{}),
+	}
+	return s
+
+}
+
+// Calling Start will set the RuleIter, unblock the MemStore, and start the run() function in a separate goroutine.
+func (m *MemStore) Start(iter RuleIter) {
+	m.mgr = iter
+	close(m.initiated)
+	go m.run()
+}
+
+func (m *MemStore) Stop() {
+	select {
+	case <-m.initiated:
+	default:
+		// If initiated is blocked, the MemStore has yet to start: easy no-op.
+		return
+	}
+
+	// Need to nil all series & decrement gauges
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	select {
+	// ensures Stop() is idempotent
+	case <-m.done:
+		return
+	default:
```

Comment:
similar concern, I think both of these could be executed, not sure this guarantees idempotency

Reply:
See above, I think this is fine.